core/src/test/scala/unit/kafka/network/ConnectionQuotasTest.scala

```diff
@@ -0,0 +1,307 @@
+      val future1 = executor.submit((() => acceptConnections(connectionQuotas, listeners("ADMIN"), 1)): Runnable)
+      intercept[TimeoutException](future1.get(1, TimeUnit.SECONDS))
+
+      // adding inter-broker connection should still succeed, even though a connection from another listener is waiting
+      executor.submit((() => acceptConnections(connectionQuotas, listeners("REPLICATION"), 1)): Runnable).get(5, TimeUnit.SECONDS)
+
+      // at this point, we need to remove 3 connections for the waiting connection to succeed
+      // remove 2 first -- should not be enough to accept the waiting connection
+      for (_ <- 0 until 2) connectionQuotas.dec(listeners("EXTERNAL").listenerName, listeners("EXTERNAL"). defaultClientIp)
+      intercept[TimeoutException](future1.get(1, TimeUnit.SECONDS))
+      connectionQuotas.dec(listeners("EXTERNAL").listenerName, listeners("EXTERNAL").defaultClientIp)
+      future1.get(1, TimeUnit.SECONDS)
+    } finally {
+      executor.shutdownNow()
+    }
+  }
+
+  @Test
+  def testMaxListenerConnectionLimits(): Unit = {
+    val maxConnections = 800
+    // sum of per-listener connection limits is below total connection limit
+    val listenerMaxConnections = 200
+    val props = brokerPropsWithDefaultConnectionLimits
+    props.put(KafkaConfig.MaxConnectionsProp, maxConnections.toString)
+    val config = KafkaConfig.fromProps(props)
+    val connectionQuotas = new ConnectionQuotas(config, time)
+
+    addListenersAndVerify(config, connectionQuotas)
+
+    val listenerConfig = Map(KafkaConfig.MaxConnectionsProp -> listenerMaxConnections.toString).asJava
+    listeners.values.foreach { listener =>
+      connectionQuotas.maxConnectionsPerListener(listener.listenerName).configure(listenerConfig)
+    }
+
+    val executor = Executors.newFixedThreadPool(listeners.size)
+    try {
+      // verify each listener can create up to max connections configured for that listener
+      val futures = listeners.values.map( listener =>
+        executor.submit((() => acceptConnections(connectionQuotas, listener, listenerMaxConnections)): Runnable) )
+      futures.foreach(_.get(5, TimeUnit.SECONDS))
+      listeners.values.foreach { listener =>
+        assertEquals(s"${listener.listenerName.value()}", listenerMaxConnections, connectionQuotas.get(listener.defaultClientIp))
+        assertFalse(s"Total number of connections on $listener should be exactly the maximum.", connectionQuotas.maxConnectionsExceeded(listener.listenerName))
+      }
+
+      // since every listener has exactly the max number of listener connections,
+      // every listener should block on the next connection creation, even the inter-broker listener
+      val futures2 = listeners.values.map( listener =>
+        executor.submit((() => acceptConnections(connectionQuotas, listener, 1)): Runnable) )
+      futures2.foreach { future =>
+        intercept[TimeoutException](future.get(1, TimeUnit.SECONDS))
```

Comment:
you may want to assert here.

Reply:
see my comment that intercept does assert on both wrong exception or no exception, similar to assertThrown